Add more Sequence Header OBU parameters to AV1CodecConfigurationRecord.

tomfinegan opened the pull request:
Add the following fields to AV1CodecConfigurationRecord:
    - version  
    - seq_profile
    - still_picture
    - seq_level_idx_0
    - seq_tier_0
    - bit_depth
    - monochrome
    - chroma_subsampling_x
    - chroma_subsampling_y
    - chrome_sample_position
    
    Also, some minor tweaks:
    - Update date.
    - Update AV1 bitstream spec status (LS -> Standard).
    - Add seq_profile, still_picture, mono_chrome to
      subsampling_x, subsampling_y, and 
      chroma_sample_position to terms defined in AV1.

tomfinegan commented:
Rendered HTML for the output diff:
https://htmlpreview.github.io/?https://github.com/AOMediaCodec/av1-isobmff/blob/c2f164deb29f674b4512c2428171413259dcad57/index.html

tomfinegan commented on index.bs:
```diff
@@ -214,12 +217,16 @@ class AV1CodecConfigurationBox extends FullBox('av1C', version = 0, 0){
 }
```
Does version need an update now?

cconcolato commented:
This is an interesting idea: 
- The part about the initial_presentation_delay should not be removed (changes to that part should be coordinated with #53)
- the fields you added are very similar to the non-optional fields required in the `codecs` parameter, except that we are missing `chromaSubsampling`. All the other fields from the `codecs` are covered with a `colr` box of subtype `nclx`. What about adding it?
- We need a sentence saying that if a SH OBU is present in configOBUs the value of those fields if present shall match the ones in the OBU (similar to the current text we have for `colr`).
- Even if the spec is only a draft, the fact that it is already implemented in many places makes it hard to make  such backwards incompatible changes, parsing-wise. I would have liked to avoid versioning but I don't see how.

fcartegnie commented:
Relying only on Fullbox versioning does not allow reusing the DCR outside of ISOBMFF mapping, unlike h26x.

tomfinegan commented:
> The part about the initial_presentation_delay should not be removed (changes to that part should be coordinated with #53)

Restored.

> the fields you added are very similar to the non-optional fields required in the codecs parameter, except that we are missing chromaSubsampling. All the other fields from the codecs are covered with a colr box of subtype nclx. What about adding it?

I dropped monochrome. I went too far with the addition of monochrome. It seemed logical to carry the same collection of values that are mandatory in the Codecs Parameter String within the AV1CodecConfigurationRecord. I can restore it and add the chromaSubsampling info as well. Do you prefer one over the other?

> We need a sentence saying that if a SH OBU is present in configOBUs the value of those fields if present shall match the ones in the OBU (similar to the current text we have for colr).

Added.

> Even if the spec is only a draft, the fact that it is already implemented in many places makes it hard to make such backwards incompatible changes, parsing-wise. I would have liked to avoid versioning but I don't see how.

Yes, same here. I've bumped it to 1.

cconcolato commented:
> Relying only on Fullbox versioning does not allow reusing the DCR outside of ISOBMFF mapping, unlike h26x.

@fcartegnie Can you elaborate? How is the DCR used outside? Why would the version and flags be a problem?



> Same remark with parameters extracted in atoms and not duplicated.

Same question. Can you give more details?

fcartegnie commented on index.bs:
```diff
@@ -212,21 +212,27 @@ The <dfn>AV1CodecConfigurationBox</dfn> contains decoder configuration informati
-class AV1CodecConfigurationBox extends FullBox('av1C', version = 0, 0){
+class AV1CodecConfigurationBox extends FullBox('av1C', version = 1, 0){
   AV1CodecConfigurationRecord av1Config;
 }
```
25 bits ?
The struct OBU array is no longer aligned to 8 bits.

tomfinegan replied:
Thanks-- Didn't account for one of the flags (initial_presentation_delay_present or sequence_parameters_present) when I adjusted the size of the first reserved field.

cconcolato commented:
@tomfinegan There is a byte alignment problem: 25 bits. 
Also the v1 applies to the box not to the config record... We should probably rename AV1CodecConfigurationRecord to AV1CodecConfigurationRecordv1 and say that the av01 v0 and AV1CodecConfigurationRecord used in v0 are deprecated.

fcartegnie commented:
> @fcartegnie Can you elaborate? How is the DCR used outside? Why would the version and flags be a problem?

For example, for h26x, some decoders have made the choice of using the DCR as initialization extradata, because the struct contains all parameters. 
But this is made possible because the DCR itself also contains versioning.

If you use DCR as extradata, you might also want to add extracted bitdepth. You then need to add atoms at the end of the struct. (plain copy of stsd from DCR)
Unsure if that's better than having the duplicate info inside DCR.

tomfinegan commented:
> I don't see why we have a sequence_parameters_present.

It allows for a simple muxer that doesn't understand how to parse SH OBUs or how to extract one for configOBUs to produce output that can be easily updated in place by an external tool.

fcartegnie commented:
> It allows for a simple muxer that doesn't understand how to parse SH OBUs or how to extract one for configOBUs to produce output that can be easily updated in place by an external tool.

This makes previous discussions about requiring profile and essential parameters effectless.
People will just use 0x0, 0x0, 0x0, and make the whole DCR useless.
All muxers and user code will need to fully understand bitstream and parse obu and have delays just to have profile and late fail with unsupported profile.

cconcolato commented on index.bs:
```diff
@@ -213,30 +218,41 @@ The <dfn>AV1CodecConfigurationBox</dfn> contains decoder configuration informati
-  unsigned int (3) reserved = 0;
+aligned (8) class AV1CodecConfigurationRecordv1 {
+  unsigned int (4) reserved = 0;
+  unsigned int (3) version;
```
As said in the text, the syntax should say `version = 1;`

tomfinegan replied:
Done.

cconcolato commented on index.bs:
```diff
@@ -248,10 +264,34 @@ The <dfn>initial_presentation_delay_minus_one</dfn> field indicates the number o
 
 When smooth presentation can be guaranteed after decoding the first sample, the value 0 SHALL be used. If an ISOBMFF writer cannot verify the above procedure, [=initial_presentation_delay_present=] SHALL be set to 0.
 
+The <dfn export>sequence_parameters_present</dfn> field indicates the presence of the parameters [=seq_profile=], [=still_picture=], [=seq_level_idx_0=], [=seq_tier_0=], and [=bit_depth=] fields from the AV1 [=Sequence Header OBU=].
```
The list is not complete anymore. I would simply say:
> indicate the presence of parameters from the [=Sequence Header OBU=], either present in the configOBUs field or in the associated samples.

jamrial commented:
I'm with @fcartegnie. At the very least Seq_profile needs to be an obligatory field.

Every other field you added (except maybe seq_level_idx_0 and bitdepth) is unnecessary since unlike the aforementioned one they have no use when choosing a decoder early during initialization, so you could remove them or leave them as optional, but without Seq_profile if you still allow no seq obus then all the efforts made here will be pointless and just leave us with a more complex av1C box. If parsers may still need to look at sample data just to get basic initialization information before a decoder can be chosen, then the current box will have virtually no benefits over the old one.

>It allows for a simple muxer that doesn't understand how to parse SH OBUs or how to extract one for configOBUs to produce output that can be easily updated in place by an external tool.

Muxers have had no issues filling the obligatory profile field of avcC. I don't understand why they would with av1C. You're taking precautions that have not been necessary before for other codecs and letting muxers to essentially leave the av1C box blank, which will give players a headache.

For that matter, may i request the first byte of this box to be distinguishable from the first header byte of a raw OBU? Something like the very first bit of the AV1CodecConfigurationRecordv1 struct being always 1 (The exact opposite of the obu_forbidden_bit in a raw obu header), or making the reserved bits be 5 instead of 4 (since type, the second field in a raw obu header, can't be 0), or directly making the reserved field 8 bits long?

robUx4 commented:
> It allows for a simple muxer that doesn't understand how to parse SH OBUs or how to extract one for configOBUs to produce output that can be easily updated in place by an external tool.

That's sounds like the first tool would produce non-compliant files and the second tool would fix it. If the first file is not made to be used as such, there is no problem here. We can still have very strict rules and allow tools to create intermediate files if that makes them happy. But dumbing down the requirements just so that the first tool can pretend to do OK file doesn't seem right.

The worst case scenario for the extra data creation is probably live streaming when you want the start to have as low latency as possible. IMO the best way to achieve this is to have as little information in the extra data as possible. That includes not having the encoder to produce codec-compliant OBUs. In other words, only relying on an internal structure.

Having a minimal extra data/AV1CodecConfigurationRecord means we can always create it beforehand. And we should only put what's necessary for proper playback. I think that means:

- seq_profile
- bit_depth
- monochrome
- color_primaries
- transfer_characteristics
- matrix_coefficients
- color_range
- subsampling_x
- subsampling_y

With just that you know your decoder and screen capabilities/requirements. Maybe signaling HDR would also be a good idea. The display capabilities can be made optional, leaving these as absolutely necessary:
- seq_profile
- bit_depth

All these data are know by the encoder user before even starting the encoder. So can be written even before a frame goes into the encoder. You will not get lower latency than this.

Having OBUs in here seemed originally to have these information at hand. But it can do another way. Then do we want to put the SH OBU in order to remove it from the bitstream where possible ? That can save some bytes but I don't know if it's worse the extra complexity.

jamrial commented:
This is being made more complex than it has any need to be. The box should something simple as

version
seq_profile
seq_level_idx_0
bit_depth
initial_presentation_delay_present
if (initial_presentation_delay_present)
  initial_presentation_delay_minus_one
configOBUs (zero or more)

Everything else is optional since a decoder *will* read all that from the sh obu. But a demuxer absolutely *needs* at the very least the profile value with no exceptions in this box, in order to choose a decoder that's guaranteed to support the stream.
